Fence Profiling Support in all backends

DavidPoliakoff opened the pull request:
So, this one needs conversation, we're really at a "design review" point.

One thing I've been promising for ages, but haven't gotten around to doing, is to add fence profiling to Kokkos. We have the callbacks (#2891), but they're currently unused.

What I see, is that we basically need a Kokkos wrapper around any fencing function. cudaDeviceSynchronize,cudaStreamSynchronize, that take in at a minimum a string name, _probably_ also a deviceID? Then we never directly call the fencing function, only the wrapper. But this change radiates out. If cudaStreamSynchronize takes a name, then so must Kokkos::Cuda::fence(), so that it can forward it to cudaStreamSynchronize. We also need compatibility, so we need a Kokkos::Cuda::fence with no args, that calls the named version.

I've sketched what this might look like in Cuda. Note that the only changes I make are to Cuda, we also need to modify generic code (CopyViews and the like), but those changes wouldn't compile.

*What I need to proceed* is for somebody to sign off on the design. Once you do that, I will write _the infrastructure for this_ in every backend. My hope is that we can merge at that point. Once we have the infrastructure, I (or a backend owner) can start labeling fences, and starting to put this in generic code as well (like that in CopyViews).

crtrott commented:
This looks reasonable to me. We need to discuss the naming scheme, and unify that accross all of Kokkos internally named things though. I know we did a lot of ClassName::function_name naming for example in messages, etc. but its not super uniform.

So basically: Kokkos::ClassName::whatever or kokkos.classname.whatever or ...

masterleinad commented:
Retest this please.

masterleinad commented:
Looks reasonable to me. I'm not quite sure how portable `deviceID` is.

DavidPoliakoff commented:
Don't rereview this yet. I'm using CI to see what all I need to convert

masterleinad commented:
The indentation checker complains. :slightly_smiling_face:

DavidPoliakoff commented:
I have to do some more technical work on it, I'm not worried about formatting until the content is ready

masterleinad commented:
Just saying that you won't see any results for the `GPU`-backends until then. :slightly_smiling_face:

dalg24 commented:
Builds have been running for ove3r 4hrs on Jenkins.  Killing them.

DavidPoliakoff commented:
I can't reproduce the hang locally. I'm going to set this branch down for a while

DavidPoliakoff commented:
Retest this please

DavidPoliakoff commented:
Retest this please

DavidPoliakoff commented:
Comments for reviewers: this PR has sprawled. Essentially, for this to be meaningful, fences had to identify what resource was being fenced. I put in a system to make that happen. Unfortunately, that system _didn't_ work for begin_parallel_for, those always just reported their device_id as being instance 0 of their given space (So CUDA was device_id_root<Cuda>). Further, some "impl_static_fence" methods forwarded to an instance fence method, which is a broken behavior, static fences and instance fences are now semantically different things. So I reworked a few ExecutionSpace implementations to differentiate static and instance fences. I also reworked `impl_instance_id` to make its answer for "what instance is this" match the answers given by fence. With all that done, everything matches up nicely

masterleinad commented on core/src/SYCL/Kokkos_SYCL.cpp:
```diff
@@ -112,14 +112,34 @@ void SYCL::print_configuration(std::ostream& s, const bool detailed) {
+      name,
+      Kokkos::Tools::Experimental::SpecialSynchronizationCases::
+          GlobalDeviceSynchronization,
+      [&]() {
+        std::lock_guard<std::mutex> lock(Impl::SYCLInternal::mutex);
```
```suggestion
      [&]() {
        // guard accessing all_queues
        std::lock_guard<std::mutex> lock(Impl::SYCLInternal::mutex);
```
and delete above.

DavidPoliakoff replied:
0168008

DavidPoliakoff commented on core/src/Kokkos_SYCL.hpp:
```diff
@@ -83,7 +83,7 @@ class SYCL {
   explicit SYCL(const sycl::queue&);
 
-  uint32_t impl_instance_id() const noexcept { return 0; }
+  uint32_t impl_instance_id() const noexcept { return 1; }
```
So, the way this is happening is that there are two classes of ExecutionSpace: those that are differentiable, and those that aren't. In the case of SYCL, there's only one instance, so impl_instace_id is set to 1. This is changing because "0" now means something special, an operation on "0" is happening on _all instances_. A sync on Cuda + 0 is a deviceSynchronize.

Also this isn't the first time impl_instance_id has been used, that gets used by Profiling already to determine which dev_id to pass to the tool

masterleinad replied:
Yes, in `SYCL` no queue is special.

masterleinad replied:
I think I'm still confused what "In the case of SYCL, there's only one instance" means to you. How is `SYCL` different from `CUDA` or `HIP` in this regard?

DavidPoliakoff replied:
Actually, looking at this more, I think SYCL is just wired differently than the others. I was looking here: https://github.com/kokkos/kokkos/blob/master/core/src/SYCL/Kokkos_SYCL_Instance.hpp#L58-L67 , and saw no constructor from a sycl::queue, which made me think all SYCL instances submit to the same queue. But that's not true, is it? It's just that in our case, the SYCL queue is stored in the SYCL instance, not the SYCLInternal. Is there a way to get a unique ID for a SYCL queue? That's probably what we need.

masterleinad replied:
Yes, an execution space can be initialized with a `sycl::queue`, see https://github.com/kokkos/kokkos/blob/7b5ffcc3224c5d4fa8ef46dd390c59f2d55752e8/core/src/Kokkos_SYCL.hpp#L84 and it is used for initializing an `SYCLInternal` object which stores a copy of it, see https://github.com/kokkos/kokkos/blob/7b5ffcc3224c5d4fa8ef46dd390c59f2d55752e8/core/src/SYCL/Kokkos_SYCL_Instance.hpp#L89. Hence, I would think that we can use the same approach as for CUDA and HIP.

DavidPoliakoff replied:
Okay. So something to the effect of "hash my SYCLInternal object ptr"? Sorry for the slow movement here, just trying to make sure we get the right thing

masterleinad replied:
Yes, AFAICT, we do the same for `CUDA` and `HIP`, i.e. the execution space classes store a pointer to `CudaInternal` resp. `HIPInternal` which contains a `cudaStream_t` resp. `hipStream_t` member.
So I would expect
```C++
uint32_t SYCL::impl_instance_id() const noexcept {
  return m_space_instance->impl_get_instance_id();
}

uint32_t SYCLInternal::impl_get_instance_id() const {
  return Kokkos::Tools::Experimental::Impl::idForInstance<Kokkos::Experimental::SYCL>(
      reinterpret_cast<uintptr_t>(this));
}
```

DavidPoliakoff commented:
@stanmoore1 and I are going to do some profiling before we merge this, but please do review for "this looks like a bad implementation" (or approve with it being a good one)

Rombur reviewed (commented):
Looks good to me so far. I only have minor comments.

DavidPoliakoff commented:
@wrwilliams @khuck:

https://github.com/kokkos/kokkos/pull/3966/files#diff-8f9a639375f85f3c1768b44ace306f545b68cd7b303dfcc4aa8df94a5b179b4bR79-R107

"Top 8 bits represent the device type. Next 7 are the device id (think GPU). Last 17 are the instance id (think stream)." It has been quite a while since I've done bitmasks, is this an idiotic way to write that? Once I have your signoff on the C++, I'll do it in C. Does C allow inline functions in headers in the same way C++ does?

DavidPoliakoff commented on core/src/impl/Kokkos_Profiling.hpp:
```diff
@@ -135,6 +137,79 @@ Kokkos_Profiling_SpaceHandle make_space_handle(const char* space_name);
+//  Kokkos::Tools::beginFence(name, 0, &handle);  // TODO: correct ID
+//  func();
+//  Kokkos::Tools::endFence(handle);
+//}
```
4c272ec

DavidPoliakoff commented on core/src/impl/Kokkos_Profiling.hpp:
```diff
@@ -135,6 +137,79 @@ Kokkos_Profiling_SpaceHandle make_space_handle(const char* space_name);
+    auto ret         = offset + value++;
+    (*map)[instance] = ret;
+    return ret;
+    // find->second = (offset + value++);
```
4c272ec

DavidPoliakoff commented on core/src/Threads/Kokkos_ThreadsExec.cpp:
```diff
@@ -289,20 +289,43 @@ int ThreadsExec::in_parallel() {
+        fence_lam);
+  } else {
+    Kokkos::Tools::Experimental::Impl::profile_fence_event<Kokkos::Threads>(
+
```
4c272ec

DavidPoliakoff commented on core/src/Threads/Kokkos_ThreadsExec.cpp:
```diff
@@ -289,20 +289,43 @@ int ThreadsExec::in_parallel() {
+  };
+  if (is_static == Impl::fence_is_static::yes) {
+    Kokkos::Tools::Experimental::Impl::profile_fence_event<Kokkos::Threads>(
+
```
4c272ec

DavidPoliakoff commented on core/src/SYCL/Kokkos_SYCL.cpp:
```diff
@@ -112,14 +112,34 @@ void SYCL::print_configuration(std::ostream& s, const bool detailed) {
+      name,
+      Kokkos::Tools::Experimental::Impl::DirectFenceIDHandle{
+          impl_instance_id()},
+
```
4c272ec

DavidPoliakoff commented on core/src/OpenMPTarget/Kokkos_OpenMPTarget_Instance.cpp:
```diff
@@ -59,7 +59,36 @@
 namespace Experimental {
 namespace Impl {
-void OpenMPTargetInternal::fence() {}
+// DOGS
```
4c272ec

DavidPoliakoff commented on core/src/Cuda/Kokkos_Cuda_Error.hpp:
```diff
@@ -49,13 +49,21 @@
+    const std::string& name);
+void cuda_device_synchronize(const std::string& name);
+void cuda_stream_synchronize(const cudaStream_t stream,
+                             const std::string& name);
```
Think you're right. Addressed in 35f5111

masterleinad commented on core/src/Cuda/Kokkos_Cuda_Instance.cpp:
```diff
@@ -902,7 +951,15 @@ void CudaSpaceInitializer::finalize(bool all_spaces) {
+      "Kokkos::CudaSpaceInitializer::fence: Initializer Fence");
+}
+void CudaSpaceInitializer::fence(const std::string &name) {
+  // Kokkos::Cuda::impl_static_fence("Kokkos::CudaSpaceInitializer::fence:
+  // "+name); //TODO: or this
+  Kokkos::Cuda::impl_static_fence(name);
```
Is this something you want to have discussed and addressed in this pull request?

DavidPoliakoff replied:
I'll delete the comment. I basically decided on "do it the way it is now." This fence is never directly invoked, only ever invoked by another function, that function should be saying what the name of the fence is

masterleinad commented on core/src/Cuda/Kokkos_Cuda_KernelLaunch.hpp:
```diff
@@ -638,7 +638,8 @@ struct CudaParallelLaunchImpl<
       CUDA_SAFE_CALL(cudaGetLastError());
-      cuda_instance->fence();
+      cuda_instance->fence(
+          "Kokkos::Impl::launch_kernel: Debug Only Bounds Check Fence");
```
IMHO it's negotiable if `KOKKOS_ENABLE_DEBUG_BOUNDS_CHECK` is correct here instead of just `KOKKOS_ENABLE_DEBUG`. In any case, the purpose of this `fence` is to check for execution errors in the kernel just launched. Hence, I would rather say something like "Kokkos::Impl::launch_kernel: Check for execution error");".

DavidPoliakoff replied:
I want to leave "Debug Only" in there, as I want people to be able to quickly see if a fence comes from them turning on custom, non-release flags. How would you feel about

"Kokkos::Impl::launch_kernel: Debug Only Check for Execution Error"?

masterleinad replied:
Sure, that sounds reasonable.

DavidPoliakoff replied:
552bb3c

masterleinad commented on core/src/HIP/Kokkos_HIP_Instance.cpp:
```diff
@@ -154,6 +154,10 @@ int HIPInternal::verify_is_initialized(const char *const label) const {
 }
 
+uint32_t HIPInternal::impl_get_instance_id() const noexcept {
+  return Kokkos::Tools::Experimental::Impl::idForInstance<
+      Kokkos::Experimental::HIP>(reinterpret_cast<uintptr_t>(this));
```
Could you please elaborate on why you need all this magic here? In which cases do you want the same id to be returned? Does this work the same way for all backends?

DavidPoliakoff replied:
`Does this work the same way for all backends?` For all backends which have distinct instances, it works this way. Serial doesn't work this way, because it only has the one instance.

I need all of this magic because for distinguishable instances, I need to map those instances to an ID. I want the same ID to be returned if the same instance is used. So if somebody creates a CUDA/HIP instance from different streams, we would want to see different ID's here.

masterleinad replied:
I was slightly confused that using the address to create a hash is what we want. Of course, this works because these classes are always stored as pointers in the execution space classes.

masterleinad commented on core/src/Kokkos_Parallel.hpp:
```diff
@@ -175,14 +175,16 @@ template <class ExecPolicy, class FunctorType>
 
   ::Kokkos::parallel_for(policy, functor, str);
 
 #if KOKKOS_ENABLE_DEBUG_PRINT_KERNEL_NAMES
```
Looks like this is never defined...

DavidPoliakoff replied:
@crtrott can correct me, I believe this is an advanced feature, only set in the CMAKE_CXX_FLAGS, there's no corresponding CMake option

masterleinad replied:
Yes, that's what I meant. Obviously, not related to this pull request. Maybe, we discuss somewhere else if we should document this better.

DavidPoliakoff replied:
Of course, such users *should* be using the Kokkos Tools subsystem, writen by that smrt guy, for these porpoises

masterleinad replied:
See https://github.com/kokkos/kokkos/pull/4150 for discussing if/what we want to do about it.

masterleinad commented on core/src/Kokkos_Serial.hpp:
```diff
@@ -120,9 +120,24 @@ class Serial {
+        name,
+        Kokkos::Tools::Experimental::SpecialSynchronizationCases::
+            GlobalDeviceSynchronization,
+        []() {});  // TODO: correct device ID
```
What does "device ID" mean for `Serial`?

DavidPoliakoff replied:
Here I should just delete the comment, it's out-of-date. For Serial, you'll only see two device ID's, The first is the one you'll hit here. This is hit when somebody calls Kokkos::fence, and it's a fence of "the Serial Device." It's the semantic equivalent of a cudaDeviceSynchronize. The ID we provide there will be [the Serial DeviceType ID] << 24 (and as Serial = 0, this will be 0). The other will be for an instance fence. Now, in the Serial case, there's only one distinguishable instance, there's no "create a Serial instance from a Serial stream." So the TL;DR is that there will be only two instances: 0 and 1, for Serial.

crtrott replied:
That might actually change. If you have external threads, for thread safety we may need to have multiple instances for example to have individual scratch pads.

masterleinad reviewed (commented):
I have some more suggestions, questions, and comments but overall this looks reasonable to me.

wrwilliams commented on core/src/impl/Kokkos_Profiling_Interface.hpp:
```diff
@@ -68,15 +76,49 @@ enum struct DeviceType {
+struct ExecutionSpaceIdentifier {
+  DeviceType type;
+  uint32_t device_id;
+  uint32_t instance_id;
```
`instance_id` here is theoretically 1:1 with, but not necessarily equal to, a CUDA/HIP stream ID, right? Is that mapping exposed somewhere?

DavidPoliakoff replied:
@wrwilliams, in the 99% case, you're correct that it's 1:1. A user _technically_ could create multiple distinct Kokkos::[Cuda/HIP] instances from the same stream. There's no valid reason to, but a person _could_.

That mapping is not exposed anywhere as of yet, I think a tool would need to discover it (something like monitor CUDA and see what stream a kernel was enqueued on). I could look into a scheme for this in a follow-on PR

DavidPoliakoff commented on core/src/impl/Kokkos_Profiling.hpp:
```diff
@@ -135,6 +137,69 @@ Kokkos_Profiling_SpaceHandle make_space_handle(const char* space_name);
+
+  auto find = map->find(instance);
+  if (find == map->end()) {
+    auto ret         = offset + value++;
```
In time, probably? But as of now I don't think we're requiring thread-safety.

crtrott replied:
actually this needs to be thread safe right now since you call this without a tool being loaded, and stuff like the CUDA backend is used in a threaded mode already.

DavidPoliakoff commented on core/src/impl/Kokkos_Profiling.hpp:
```diff
@@ -135,6 +137,69 @@ Kokkos_Profiling_SpaceHandle make_space_handle(const char* space_name);
+  /** Needed to be a ptr due to initialization order problems*/
+  using map_type = std::map<uintptr_t, uint32_t>;
+
+  static std::shared_ptr<map_type> map;
+  if (map == nullptr) {
+    map = std::make_shared<map_type>();
+  }
```
I'd have to look back pretty far into the CI. Back when it was a map_type, I recall hitting segfaults in Clang builds

DavidPoliakoff replied:
I'll push with just a map_type, and see if it passes. That could have been an artifact of earlier builds

DavidPoliakoff replied:
Oddly, this passes. I must have done something else goofy. My bad, I like this version a lot better.

DavidPoliakoff replied:
Whoops, it might fail (https://github.com/kokkos/kokkos/pull/3966/checks?check_run_id=3021880153 ). I'll try reverting to a shared_ptr

DavidPoliakoff commented on core/src/impl/Kokkos_Profiling.hpp:
```diff
@@ -135,6 +137,69 @@ Kokkos_Profiling_SpaceHandle make_space_handle(const char* space_name);
+    map = std::make_shared<map_type>();
+  }
+
+  static uint32_t value;
```
It really isn't necessary, statics are defined to be zero-initialized in the standard. Personally I don't like how it reads, it has always been counterintuitive to me that the initializer for a static is only run the first time the function is encountered. Not explicitly calling for initialization always looks better to me. It's a personal preference thing, though, happy to have another voice say "explicitly initialize" and I'll follow that guidance

DavidPoliakoff commented on core/src/Cuda/Kokkos_Cuda_Error.hpp:
```diff
@@ -49,13 +49,19 @@
 namespace Impl {
 
-void cuda_device_synchronize();
+void cuda_stream_synchronize(
+    const cudaStream_t stream,
+    Kokkos::Tools::Experimental::SpecialSynchronizationCases reason,
+    const std::string& name);
+void cuda_device_synchronize(const std::string& name);
+void cuda_stream_synchronize(const cudaStream_t stream,
+                             const std::string& name);
```
Yes, the overloads do entirely different things. One implements a case where a user calls cuda_stream_sync directly, the other is for cases where a special sync is happening (like on the deep copy stream, which doesn't have an associated Cuda instance to look up)

crtrott commented on core/src/Cuda/Kokkos_Cuda_Instance.cpp:
```diff
@@ -141,7 +141,36 @@ bool cuda_launch_blocking() {
+  Kokkos::Tools::Experimental::Impl::profile_fence_event<Kokkos::Cuda>(
+      name,
+      Kokkos::Tools::Experimental::Impl::DirectFenceIDHandle{
+          ptr->impl_get_instance_id()},
```
I am slightly unhappy that this kind of work (looking up ID's etc.) happens even if no profiling tool is loaded, but the benchmarks I ran don't seem to indicate too much impact. Back to back fences are now about 20% slower, but that is from a very low base of about half a microsecond on CUDA. having a empty kernel in between we loose roughly the same absolute time of 0.1us but on a total time of 5us. So we are talking about 2% slow down for a pure cached latency benchmark. What I don't know is whether the time goes up significantly if the underlying map is not in cache anymore. On the other hand if its not in cache anymore that means you moved enough data to evict it, so the time between fences should be substantially larger, reducing the overhead impact.

DavidPoliakoff replied:
So, I have a way around this, if you'd like to see it explored. Essentially, rather than pass the value, I could pass a function that returns the value, and only invoke the function if a tool is loaded. My suspicion is that this is getting a little baroque for these purposes, but if you want me to check that out I can

DavidPoliakoff replied:
Here's where I've landed on this: I think we should unify all of the "only do this if a tool is loaded" stuff into a new interface that I have in mind. If we don't think this overhead is catastophic, I'd recommend we merge as is, and I'll take another stab at basically a "lazily acquire callback values" system.

crtrott commented on core/src/Kokkos_Threads.hpp:
```diff
@@ -107,8 +108,13 @@ class Threads {
   static void impl_static_fence();
+  static void impl_static_fence(const std::string& name);
+
+  void fence(Impl::fence_is_static is_static) const;
```
what are these new overloads, and why are there public functions which only exist for this backend?

DavidPoliakoff replied:
These functions exist for any backend on which there was no separation between a "static" fence and an "instance" fence. In the past, an instance fence was the same thing as a static fence, in terms of implementation. The ugly part of this PR is that while these are still mechanically the same thing, we need to be able to differentiate them for profiling. So I basically just wrote these to provide the plumbing I needed to tell the tools subsystem about a "Whole Device" fence or "instance fence." These still mechanically do the same thing, except in their reporting.

I could probably make the instance fence that takes in `is_static` into an impl method, but I was kinda leaning on "functions with Impl args are Impl." But if we want to make the function itself impl, I'm not too hostile to that

crtrott replied:
I would feel better with that then being an internal function. We already distinguish between the static fence and the non-static fence technically: impl_static_fence and fence, with the static version only being there for internal purposes such as for the implementation of Kokkos::fence(); So if the reason right now is that impl_static_fence() calls fence(), then just make them both call somethign internal.

crtrott commented on core/src/OpenMPTarget/Kokkos_OpenMPTarget_Instance.hpp:
```diff
@@ -51,14 +51,18 @@ namespace Kokkos {
-  void fence();
+  void fence(openmp_fence_is_static is_static = openmp_fence_is_static::no);
+  void fence(const std::string& name,
+             openmp_fence_is_static is_static = openmp_fence_is_static::no);
```
same here as for threads, this is a new public overload. Might be better to delegate to an impl_funciton

DavidPoliakoff replied:
See: https://github.com/kokkos/kokkos/pull/3966#discussion_r666305635

crtrott reviewed (changes requested):
All in all really good PR. 

Couple points: 
1) I am slightly concerned about adding work to fences even if no tool is loaded (like ID lookup) but my own benchmarks seems to indicate its fairly minimal (see my comment) so I won't block on it. 
2) I think the overloads of fence for OpenMP/THreads with the "is static" enum should not be public. Though I guess with the enums being typed and not public that is enough?
3) I am wondering about information overload with the very fine grained naming we got here. For somebody intimately familiar with Kokkos its really cool, but I wonder if the average developer is overwhelmed. Also in some cases I think we should not literally use the Impl detail name, like for Kokkos::Impl::DeepCopy. I wonder if its better to use the user facing calling entity here Kokkos::deep_copy, since that is what is causing this ultimately and might be easier for users to connect to their code?

crtrott commented:
Note the request changes is more a "request short discussion of my points" I am basically ok with this PR.

masterleinad reviewed (changes requested):
I'm happy with this pull request if you treat `SYCL` the same as `HIP` and `CUDA` wrt. the instance id.

crtrott reviewed (changes requested):
I forget to answer that thing you guys resolved earlier about thread safety: I fear we need this to be thread safe.

masterleinad commented on core/src/impl/Kokkos_Profiling.hpp:
```diff
@@ -144,6 +145,7 @@ struct DirectFenceIDHandle {
 //
 template <typename Space>
 uint32_t idForInstance(const uintptr_t instance) {
+  const std::lock_guard<std::mutex> lock(std::mutex{});
```
Doesn't this create a local variable of type std::mutex that is different for every functional call? Shouldn't this be something like
```suggestion
  static std::mutex instance_mutex;
  const std::lock_guard<std::mutex> lock(instance_mutex);
```

DavidPoliakoff replied:
Oh right. Slow moment, thanks

DavidPoliakoff replied:
a51ebe9

DavidPoliakoff commented:
@crtrott: how do you feel about this as a resolution for thread-safety? Basically we're only accessing the map on ExecSpace construction. This could still be a bit slower when a bunch of threads are making spaces, but I honestly think that ID lookup will be a lot cheaper than all of the scratch allocations and the like

DavidPoliakoff commented:
@crtrott: let me know if you want further edits (like making those functions Impl). Personally I'm okay with this PR as-is

crtrott reviewed (changes requested):
Make the new overloads for fence internal functions instead, in particular since they don't exist for every execution space.

dalg24 reviewed (commented):
(Not a thorough review)

DavidPoliakoff commented:
Good catches, I'll fix those

DavidPoliakoff commented on core/src/Cuda/Kokkos_CudaSpace.cpp:
```diff
@@ -405,7 +412,8 @@ void CudaUVMSpace::impl_deallocate(
   }
-  Cuda::impl_static_fence();
+  Cuda::impl_static_fence(
+      "Kokkos::CudaUVMSpace::impl_allocate: Post UVM Deallocation");
```
4123689

DavidPoliakoff commented on core/src/Cuda/Kokkos_CudaSpace.cpp:
```diff
@@ -391,7 +397,8 @@ void CudaUVMSpace::impl_deallocate(
     const Kokkos::Tools::SpaceHandle arg_handle) const {
-  Cuda::impl_static_fence();
+  Cuda::impl_static_fence(
+      "Kokkos::CudaUVMSpace::impl_allocate: Pre UVM Deallocation");
```
4123689
